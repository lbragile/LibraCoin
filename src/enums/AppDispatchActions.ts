export enum ACTIONS {
  ADD_VERIFIED_TRANS,
  UPDATE_VERIFIED_TRANS,
  UPDATE_SELECTED_TRANS,
  UPDATE_USERS,
  SET_MAIN_USER,
  ADD_BLOCK,
  UPDATE_BLOCK
}
