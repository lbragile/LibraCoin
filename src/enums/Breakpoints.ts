// https://www.w3schools.com/css/css_rwd_mediaqueries.asp

export enum SIZES {
  xs = "600px",
  sm = "600px",
  md = "768px",
  lg = "992px",
  xl = "1200px"
}

export enum DEVICES {
  xs = "max-width: " + SIZES.xs,
  sm = "min-width: " + SIZES.sm,
  md = "min-width: " + SIZES.md,
  lg = "min-width: " + SIZES.lg,
  xl = "min-width: " + SIZES.xl
}
