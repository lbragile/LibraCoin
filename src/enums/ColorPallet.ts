export enum COLORS {
  DEFAULT_BACKGROUND = "#fff",

  DARK_BACKGROUND = "#1f1f1f",

  VALID_BACKGROUND = "#b4ff9a",
  INVALID_BACKGROUND = "#ff9a9a",
  VALID_SOLUTION = "#008000",
  INVALID_SOLUTION = "#f00",

  NAVBAR_BACKGROUND = "#eee",
  NAV_LINK_UNDERSCORE = "#8ac0ff",
  NAV_TEXT = "#000",

  SELECTED_TRANS_BACKGROUND = "#dfd",
  SELECTED_TRANS_BORDER = "#090",

  BLOCK_TRANS_BACKGROUND = "#c7e4ff",

  ROOT_CELL = "#4f9249",
  MIDDLE_CELLS = "#cf8b31",
  DATA_CELLS = "#7d95ff",
  PREVIEW_LINES = "#777"
}
