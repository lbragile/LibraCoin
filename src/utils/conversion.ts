import { TextEncoder } from "text-encoding";

export function bufferToHex(hashBuffer: ArrayBuffer): string {
  const hashArray = Array.from(new Uint8Array(hashBuffer)); // convert buffer to byte array
  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join(""); // convert bytes to hex string
  return hashHex;
}

export async function CryptoKeyToHex(format: string, key: CryptoKey): Promise<string> {
  const buf = (await window.crypto.subtle.exportKey(format, key)) as ArrayBuffer;
  return bufferToHex(buf);
}

export async function digestMessage(message: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
  const hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8); // hash the message
  return bufferToHex(hashBuffer);
}

export function randomHash(len: number): string {
  return bufferToHex(window.crypto.getRandomValues(new Uint32Array(len)));
}
