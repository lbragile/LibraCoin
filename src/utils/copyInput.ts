export function copyKey(
  e: React.FocusEvent<HTMLTextAreaElement | HTMLInputElement>,
  setCopied: (arg: boolean[]) => void,
  type?: "public" | "private"
): void {
  e.target.select();
  e.target.setSelectionRange(0, 1e6);
  document.execCommand("copy");

  // wallet page, copying keys
  if (type) {
    const isPublic = type === "public";
    setCopied([isPublic, !isPublic && !e.target.value.includes("â—¦")]);
  } else {
    setCopied([true]);
  }
}
