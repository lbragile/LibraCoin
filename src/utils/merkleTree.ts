import { Tree } from "./Tree";
import { ITransaction } from "../typings/AppTypes";
import { digestMessage } from "./conversion";

export async function calculateMerkleTreeFormation(
  verifiedTrans: ITransaction[],
  selectedTrans: ITransaction[],
  setMerkleTree?: (arg: string[][]) => void
): Promise<string> {
  let tree = [[""]];
  if (selectedTrans.length > 0) {
    // need to make sure node's in tree appear in same order as in the verified transaction pane, regardless of selection order
    const verifiedSignatures = verifiedTrans.map((trans) => trans.signature);
    const selectedSignatures = selectedTrans.map((trans) => trans.signature);
    let signatures = verifiedSignatures.filter((sig) => selectedSignatures.includes(sig));
    tree = [signatures];

    while (signatures.length !== 1) {
      const hashArr = [] as string[];
      for (let i = 0; i < signatures.length; i += 2) {
        const hash = signatures[i + 1] ? await digestMessage(signatures[i] + signatures[i + 1]) : signatures[i];
        hashArr.push(hash);
      }

      signatures = hashArr;
      tree.push(hashArr);
    }
  }

  if (setMerkleTree) {
    setMerkleTree(tree);
  }

  return getMerkleRoot(tree);
}

export function getMerkleRoot(tree: string[][]): string {
  return tree[tree.length - 1][0];
}

export function flattenTree(tree: string[][]): string[] {
  let flatTree = [] as string[];
  // need to reverse each tree row so that printing produces the right order!
  for (let i = 0; i < tree.length; i++) {
    flatTree = flatTree.concat(tree[i].reverse());
  }

  return flatTree;
}

export function drawTreeDiagramOnCanvas(
  merkleTree: string[][],
  canvas: HTMLCanvasElement | null,
  transactions: ITransaction[]
): void {
  if (canvas) {
    const canvasTree = new Tree(canvas, transactions);
    canvasTree.clear();
    const flatTree = flattenTree(merkleTree);
    for (let i = flatTree.length - 1; i >= 0; i--) {
      canvasTree.addNode(flatTree[i]);
    }
    canvasTree.drawTree();
  }
}
