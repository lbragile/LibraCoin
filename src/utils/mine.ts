import { digestMessage, randomHash } from "./conversion";

async function createTarget(numZeros: number): Promise<string> {
  const targetHash = await digestMessage(randomHash(20));

  // replace leading bits with zeros
  const re = new RegExp(`^.{0,${numZeros}}`, "g");
  const zerosStr = Array(numZeros).fill("0").join("");

  return targetHash.replace(re, zerosStr);
}

export async function mine(
  nonce: number,
  setHeader: (arg: number) => void,
  setTarget: (arg: string) => void,
  setSolution: (arg: string) => void,
  setIsValid: (arg: boolean) => void
): Promise<void> {
  // make target
  const numZeros = Math.round(Math.random()) + 2;
  const targetHash = await createTarget(numZeros);
  setTarget(targetHash);

  // mine
  let candidateSolution = "";
  let header = nonce; // re-assigning for clarity
  while (header <= Number.MAX_SAFE_INTEGER) {
    candidateSolution = await digestMessage(header.toString());
    setSolution(candidateSolution);
    setHeader(header++);

    const leadingBits = candidateSolution.substr(0, numZeros).split("");
    if (leadingBits.every((bit) => bit === "0")) {
      break;
    }
  }

  setIsValid(candidateSolution <= targetHash);
}
