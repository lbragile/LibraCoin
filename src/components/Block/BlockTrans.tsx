import React, { useState } from "react";
import { Form, InputGroup } from "react-bootstrap";
import { IBlock, ITransaction } from "../../typings/AppTypes";
import { digestMessage } from "../../utils/conversion";
import { calculateMerkleTreeFormation } from "../../utils/merkleTree";

import "./Block.css";

type TChange = "from" | "to" | "message" | "amount";

export default function BlockTrans({
  block,
  setMerkleRoot,
}: {
  block: IBlock;
  setMerkleRoot: (arg: string) => void;
}): JSX.Element {
  const [trans, setTrans] = useState<ITransaction[]>(block.transactions);

  async function calculateNewMerkleRoot(
    e: React.ChangeEvent<HTMLInputElement>,
    index: number,
    type: TChange
  ): Promise<void> {
    const newVal = e.target.value;
    const newTrans = JSON.parse(JSON.stringify(trans)); // deep copy

    // update the changed value
    newTrans[index] = { ...newTrans[index], [type]: newVal };

    // new signature
    newTrans[index].signature = await digestMessage(newTrans[index].to + newTrans[index].from + newTrans[index].amount + newTrans[index].message); // prettier-ignore
    setTrans(newTrans);

    const root = await calculateMerkleTreeFormation(newTrans, newTrans);
    setMerkleRoot(root);

    // TODO: persist changes, and propagate details to blocks in Block.tsx
  }

  return (
    <div className="bg-dark col p-1 rounded">
      {trans.map((transaction: ITransaction, i: number) => {
        return (
          <div className="bg-light p-2 m-1 rounded" key={Math.random()}>
            <Form.Group className="mb-2 text-center">
              <Form.Control
                type="text"
                value={transaction.from}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => calculateNewMerkleRoot(e, i, "from")}
              />
              <h3 className="my-0">â†“</h3>
              <Form.Control
                type="text"
                value={transaction.to}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => calculateNewMerkleRoot(e, i, "to")}
              />
            </Form.Group>

            <InputGroup className="mb-2">
              <InputGroup.Prepend>
                <InputGroup.Text>Msg</InputGroup.Text>
              </InputGroup.Prepend>
              <Form.Control
                as="textarea"
                value={transaction.message}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => calculateNewMerkleRoot(e, i, "message")}
              />
            </InputGroup>

            <InputGroup className="mb-2">
              <Form.Control
                type="number"
                value={transaction.amount && parseFloat(transaction.amount + "").toFixed(2)}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) => calculateNewMerkleRoot(e, i, "amount")}
              />
              <InputGroup.Append>
                <InputGroup.Text>LC</InputGroup.Text>
              </InputGroup.Append>
            </InputGroup>

            <InputGroup className="mb-2">
              <InputGroup.Prepend>
                <InputGroup.Text>Sig</InputGroup.Text>
              </InputGroup.Prepend>
              <Form.Control type="text" defaultValue={transaction.signature} disabled={true} />
            </InputGroup>
          </div>
        );
      })}
    </div>
  );
}
