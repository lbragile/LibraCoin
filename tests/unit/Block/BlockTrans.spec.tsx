/**
 * @group unit
 */

import React from "react";
import { screen, waitFor } from "@testing-library/react";

import * as ConversionUtil from "../../../src/utils/conversion";

import BlockTrans from "../../../src/components/Block/BlockTrans";
import { customRender } from "../../utils/testUtils";
import userEvent from "@testing-library/user-event";
import { ACTIONS } from "../../../src/enums/AppDispatchActions";

const { initialState } = global;

it("renders correctly", async () => {
  const { asFragment } = customRender(<BlockTrans index={2} />);

  const form = await screen.findByRole("listitem", { name: /Block Transactions Item/i });
  expect(form).toBeVisible();
  expect(form).toHaveFormValues({
    btFrom: "Z",
    btTo: "A",
    btMsg: "Forth Transaction",
    btAmount: 567.89,
    btSignature: "AZ567.89"
  });

  expect(asFragment()).toMatchSnapshot();
});

describe("middle block", () => {
  const dispatchMock = jest.fn();

  // make only second block show the transactions
  const newInitialState = JSON.parse(JSON.stringify(initialState));
  newInitialState.chain[1].showTrans = true;
  newInitialState.chain[2].showTrans = false;

  beforeEach(() => {
    jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValue("some message");
    Date.now = jest.fn().mockReturnValue(12345);
    dispatchMock.mockClear();
  });

  test.each`
    field        | transaction
    ${"From"}    | ${{ amount: 987.65, from: "E1", msg: "Forth Transaction", signature: "some message", to: "A" }}
    ${"To"}      | ${{ amount: 987.65, from: "E", msg: "Forth Transaction", signature: "some message", to: "A1" }}
    ${"Amount"}  | ${{ amount: "987.651", from: "E", msg: "Forth Transaction", signature: "some message", to: "A" }}
    ${"Message"} | ${{ amount: 987.65, from: "E", msg: "Forth Transaction1", signature: "some message", to: "A" }}
  `("change of $field input causes propagation", async ({ field, transaction }) => {
    const { asFragment } = customRender(<BlockTrans index={1} />, { dispatchMock, stateMock: newInitialState });

    expect(await screen.findByRole("listitem", { name: /Block Transactions Item/i })).toBeInTheDocument();

    const name = new RegExp(`Block Transactions ${field}`, "i");
    userEvent.type(screen.getByRole(field === "Amount" ? "spinbutton" : "textbox", { name }), "1");

    // check that signature changes
    await waitFor(() => expect(dispatchMock).toHaveBeenCalledTimes(1));

    expect(dispatchMock).toHaveBeenCalledWith({
      type: ACTIONS.UPDATE_BLOCK,
      payload: {
        block: [
          {
            index: 1,
            timestamp: 12345,
            prevHash: "0000000000000000000000000000000000000000000000000000000000000000",
            currHash: "some message",
            transactions: [transaction],
            merkleRoot: "some message",
            showTrans: true,
            valid: false
          },
          {
            index: 2,
            timestamp: 12345,
            prevHash: "some message",
            currHash: "some message",
            transactions: [{ amount: 567.89, from: "Z", msg: "Forth Transaction", signature: "AZ567.89", to: "A" }],
            merkleRoot: "567.89ZForthTransactionAZ567.89A",
            showTrans: false,
            valid: false
          }
        ]
      }
    });

    expect(screen.getByRole("textbox", { name: /Block Transactions Signature/i })).toHaveValue("some message");

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("last block", () => {
  const dispatchMock = jest.fn();

  beforeEach(() => {
    jest.spyOn(ConversionUtil, "digestMessage").mockResolvedValue("some message");
    Date.now = jest.fn().mockReturnValue(12345);
    dispatchMock.mockClear();
  });

  test.each`
    field        | transaction
    ${"From"}    | ${{ amount: 567.89, from: "Z1", msg: "Forth Transaction", signature: "some message", to: "A" }}
    ${"To"}      | ${{ amount: 567.89, from: "Z", msg: "Forth Transaction", signature: "some message", to: "A1" }}
    ${"Amount"}  | ${{ amount: "567.891", from: "Z", msg: "Forth Transaction", signature: "some message", to: "A" }}
    ${"Message"} | ${{ amount: 567.89, from: "Z", msg: "Forth Transaction1", signature: "some message", to: "A" }}
  `("change of $field input attempts propagation", async ({ field, transaction }) => {
    const { asFragment } = customRender(<BlockTrans index={2} />, { dispatchMock });

    expect(await screen.findByRole("listitem", { name: /Block Transactions Item/i })).toBeInTheDocument();

    const name = new RegExp(`Block Transactions ${field}`, "i");
    userEvent.type(screen.getByRole(field === "Amount" ? "spinbutton" : "textbox", { name }), "1");

    // check that signature changes
    await waitFor(() => expect(dispatchMock).toHaveBeenCalledTimes(1));

    expect(dispatchMock).toHaveBeenCalledWith({
      type: ACTIONS.UPDATE_BLOCK,
      payload: {
        block: [
          {
            index: 2,
            timestamp: 12345,
            prevHash: "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
            currHash: "some message",
            transactions: [transaction],
            merkleRoot: "some message",
            showTrans: true,
            valid: false
          }
        ]
      }
    });

    expect(screen.getByRole("textbox", { name: /Block Transactions Signature/i })).toHaveValue("some message");

    expect(asFragment()).toMatchSnapshot();
  });
});
