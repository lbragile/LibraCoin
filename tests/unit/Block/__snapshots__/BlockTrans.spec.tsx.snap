// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last block change of Amount input attempts propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="Z"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="567.89"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`last block change of From input attempts propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="Z1"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="567.89"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`last block change of Message input attempts propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="Z"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction1
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="567.89"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`last block change of To input attempts propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="Z"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A1"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="567.89"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`middle block change of Amount input causes propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="E"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="987.65"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`middle block change of From input causes propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="E1"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="987.65"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`middle block change of Message input causes propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="E"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction1
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="987.65"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`middle block change of To input causes propagation 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="E"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A1"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="987.65"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="some message"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`renders correctly 1`] = `
<DocumentFragment>
  .c0.c0.c0 {
  font-weight: bold;
}

<div
    aria-label="Block Transactions Group"
    class="row flex-nowrap overflow-auto mx-1 mt-3 p-2 rounded bg-secondary"
    role="list"
  >
    <form
      aria-label="Block Transactions Item"
      class="col-12 bg-light p-2 rounded"
      role="listitem"
    >
      <div
        class="mb-2 text-center input-group"
      >
        <input
          aria-label="Block Transactions From"
          class="text-truncate rounded w-100 form-control"
          name="btFrom"
          type="text"
          value="Z"
        />
        <h3
          class="my-0 w-100"
        >
          ↓
        </h3>
        <input
          aria-label="Block Transactions To"
          class="text-truncate rounded w-100 form-control"
          name="btTo"
          type="text"
          value="A"
        />
      </div>
      <div
        class="mb-2 input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Msg
          </span>
        </div>
        <textarea
          aria-label="Block Transactions Message"
          class="form-control"
          name="btMsg"
          rows="3"
        >
          Forth Transaction
        </textarea>
      </div>
      <div
        class="mb-2 input-group"
      >
        <input
          aria-label="Block Transactions Amount"
          class="form-control"
          name="btAmount"
          type="number"
          value="567.89"
        />
        <div
          class="input-group-append"
        >
          <span
            class="c0 input-group-text"
          >
            LC
          </span>
        </div>
      </div>
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="c0 input-group-text"
          >
            Sig
          </span>
        </div>
        <input
          aria-label="Block Transactions Signature"
          class="text-truncate form-control"
          name="btSignature"
          readonly=""
          type="text"
          value="AZ567.89"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
